app/models/manageiq/providers/redhat/infra_manager/admin_ui.rb

```diff
@@ -0,0 +1,96 @@
+module ManageIQ::Providers::Redhat::InfraManager::AdminUI
+  extend ActiveSupport::Concern
+
+  require 'openssl'
+  require 'ovirtsdk4'
+  require 'uri'
+  require 'json'
+
+  def queue_generate_admin_ui_url
+    task_opts = {
+      :action => "Generate oVirt Admin UI URL for EMS"
+    }
+
+    queue_opts = {
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => "generate_admin_ui_url",
+      :priority    => MiqQueue::HIGH_PRIORITY,
+      :role        => "ems_operations",
+      :zone        => my_zone,
+      :args        => []
+    }
+
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def generate_admin_ui_url
+    $rhevm_log.info("Generating oVirt Admin UI URL for EMS with identifier '#{id}'.")
+
+    # Get the credentials of the provider:
+    user = authentication_userid(:default)
+    password = authentication_password(:default)
+
+    # Create the HTTP client:
+    insecure = default_endpoint.verify_ssl == OpenSSL::SSL::VERIFY_NONE,
+    ca_certs = default_endpoint.certificate_authority
+    ca_certs = nil if ca_certs.blank?
+    ca_certs = [ca_certs] if ca_certs
+    client = OvirtSDK4::HttpClient.new(
+      :insecure => insecure,
+      :ca_certs => ca_certs,
+      :log      => $rhevm_log
+    )
+
+    # Request new SSO token using the provider credentials:
+    request = OvirtSDK4::HttpRequest.new
+    request.method = :GET
+    request.url = URI::HTTPS.build(
+      :host  => default_endpoint.hostname,
+      :port  => default_endpoint.port,
+      :path  => "/ovirt-engine/sso/oauth/token",
+      :query => URI.encode_www_form(
+        :grant_type => "password",
+        :username   => user,
+        :password   => password,
+        :scope      => "ovirt-app-admin ovirt-app-portal"
+      )
+    ).to_s
+    request.headers = {
+      "Accept"     => "application/json",
+      "User-Agent" => "manageiq-providers-ovirt / API version #{api_version}"
+    }
+    client.send(request)
+    response = client.wait(request)
+
+    # Close the HTTP client:
+    client.close
+
+    # Extract the SSO token, return nil in case of error:
+    if response.code != 200
+      $rhevm_log.warn("Failed to obtain SSO token from oVirt Engine, server response follows.")
+      $rhevm_log.warn(response.message)
+      return nil
+    end
+    ssoToken = JSON.parse(response.body)['access_token']
+
+    # Generate URL to access oVirt Admin UI using the SSO token:
+    appUrl = URI::HTTPS.build(
+      :host => default_endpoint.hostname,
+      :port => default_endpoint.port,
+      :path => "/ovirt-engine/webadmin"
```

Comment:
Rubocop is going to ask you to use single quotes when there is no string interpolation or special characters: '/ovirt-engine/webadmin'.

Reply:
I'll use the single quotes as suggested, but I haven't seen any warnings from Rubocop about those so far. Anyway, what you say makes perfect sense.